controller/tests/controller_server/utils_test.py

```diff
@@ -22,32 +23,112 @@ def setUp(self):
 
+    def test_validate_secrets_with_empty_dict(self):
         secrets = {}
-        with self.assertRaises(ValidationException):
-            utils.validate_secret(secrets)
+        self._test_validate_secrets_validation_errors(secrets)
+
+    def test_validate_secrets_with_config(self):
+        secrets = {"config": str({"u1": {"username": user, "password": password, "management_address": array}}
+                                 )}
+        utils.validate_secrets(secrets)
+
+    def test_validate_secrets_with_config_invalid_secret(self):
+        secrets = {"config": str({"u1": {"username": user, "management_address": array}})}
+        self._test_validate_secrets_validation_errors(secrets)
+
+    def _test_validate_secrets_with_config_valid_uid(self, uid):
+        secrets = {"config": str({uid: {"username": user, "password": password, "management_address": array}}
+                                 )}
+        utils.validate_secrets(secrets)
+
+    def test_validate_secrets_with_config_valid_uid(self):
+        self._test_validate_secrets_with_config_valid_uid("ui_.d")
+        self._test_validate_secrets_with_config_valid_uid("a" * 90)
+
+    def test_validate_secrets_with_config_and_topologies(self):
+        secrets = {"config": str({"u1": {"username": user, "password": password, "management_address": array,
+                                         "supported_topologies": [{"topology.kubernetes.io/test": "zone1",
+                                                                   "topology.block.csi.ibm.com/test": "dev1"}]}}
+                                 )}
+        utils.validate_secrets(secrets)
+
+    def _test_validate_secrets_with_config_invalid_parameters(self, uid="u1", topology="topology.kubernetes.io/test",
+                                                              value="test"):
+        secrets = {"config": str({uid: {"username": user, "password": password, "management_address": array,
+                                        "supported_topologies": [{topology: value}]}})}
+        self._test_validate_secrets_validation_errors(secrets)
+
+    def test_validate_secrets_with_config_invalid_parameters(self):
+        uids = ["-u1", "u:1", "u1+", "u1*", "u-1(", "u/1", "u=1", " ", "", None, "a" * 91]
+        for uid in uids:
+            self._test_validate_secrets_with_config_invalid_parameters(uid=uid)
+
+        topologies = ["topology.kubernetes.io", "topology.kubernetes/test", "topology.kubernetes.io/-test"]
+        for topology in topologies:
+            self._test_validate_secrets_with_config_invalid_parameters(topology=topology)
+
+        self._test_validate_secrets_with_config_invalid_parameters(value="a*")
+
+    def _test_get_array_connection_info_from_secrets(self, secrets, topologies=None, secret_uid=None):
+        response = utils.get_array_connection_info_from_secrets(
+            secrets=secrets,
+            topologies=topologies,
+            secret_uid=secret_uid)
+        secret = response
```

Comment:
consider using something like `actual_connection_info` to begin with

Reply:
done